### PR TITLE
[vernac] Unify declaration hooks.

### PR DESCRIPTION
**Pull request by ejgallego**
Supersedes #8718.

### CODE REVIEW

**Comment by ejgallego** [comment]
Overlays:
- https://github.com/mattam82/Coq-Equations/pull/173
- https://github.com/Mtac2/Mtac2/pull/155
- https://github.com/ppedrot/ltac2/pull/106

File: doc/plugin_tutorial/tuto1/src/simple_declare.ml [hunk 1]
```diff
@@ -9,16 +9,17 @@ let edeclare ident (_, poly, _ as k) ~opaque sigma udecl body tyopt imps hook =
...
-  DeclareDef.declare_definition ident k ce ubinders imps ~hook
+  DeclareDef.declare_definition ident k ce ubinders imps ~hook_data:(hook,uctx,[])
 
 let packed_declare_definition ~poly ident value_with_constraints =
   let body, ctx = value_with_constraints in
   let sigma = Evd.from_ctx ctx in
   let k = (Decl_kinds.Global, poly, Decl_kinds.Definition) in
   let udecl = UState.default_univ_decl in
-  let nohook = Lemmas.mk_hook (fun _ x -> ()) in
+  let nohook = Lemmas.mk_hook (fun _ _ _ x -> ()) in
```
Comment (SkySkimmer) [inline doc/plugin_tutorial/tuto1/src/simple_declare.ml]: Should we keep messing around with empty hooks in the plugin tutorial? I'm inclined to just remove the hook argument to edeclare.

Comment (ejgallego) [reply doc/plugin_tutorial/tuto1/src/simple_declare.ml]: I was about to make it optional, but I hesitated due to the fact that people may forget them. As you wish.

Comment (ejgallego) [reply doc/plugin_tutorial/tuto1/src/simple_declare.ml]: Anyways if everything goes well we will greatly refactor hooks / terminators so this won't stay for long.

File: plugins/funind/indfun_common.ml [hunk 2]
```diff
@@ -144,7 +144,9 @@ let save with_clean id const ?hook (locality,_,kind) =
 	(locality, ConstRef kn)
   in
   if with_clean then Proof_global.discard_current ();
-  Lemmas.call_hook ?hook ~fix_exn l r;
+  (* XXX: need help *)
+  let uctx = UState.empty in
```
Comment (SkySkimmer) [inline plugins/funind/indfun_common.ml]: Make the `save` function take a ustate? Or make the caller (there's only 1) call the hook themselves (ie 6e8ee9cd4c3da1d743664c994f7c61d2009926c4)

Comment (ejgallego) [reply plugins/funind/indfun_common.ml]: The whole `save` function should be indeed removed I think; but let's add the argument for now.

**Comment by ejgallego** [comment]
Ready to go, any of the vernac maintainers may assign @mattam82 @maximedenes

File: vernac/lemmas.mli [hunk 3]
```diff
@@ -13,23 +13,34 @@ open Decl_kinds
...
+ * etc...
+ *
+ * Here, we use an extended hook type suitable for obligations /
+ * equations.
+ *)
+(** [hook_type] passes to the client:
+    - [ustate]: universe constraints obtained when the term was closed
+    - [(n1,t1),...(nm,tm)]: XXXX
+    - [locality]: Locality of the original declaration
+    - [ref]: identifier of the origianl declaration
```
Comment (mattam82) [inline vernac/lemmas.mli]: [(n1, t1)] .. association list between obligation name and the corresponding defined term (might be a constant, but also an arbitrary term in the Expand case of obligations)
[ref] global reference of the declared definition

The first mapping could be generalized to a mapping between goal id (however we chose to encode it) and the defined constants corresponding to it (e.g. if Qed were to define all the initial goals as constants for example)

**Review by mattam82 — APPROVED** [review]
I like where this is going, the uniformity looks good too.

File: vernac/lemmas.mli (hunk 3 above)
Comment (ejgallego) [reply vernac/lemmas.mli]: Thanks @mattam82 , updating. Indeed I think this can be made general-purpose.

**Comment by ejgallego** [comment]
> I like where this is going, the uniformity looks good too.

Thanks for the review; rebased and ready to merge.

I think I'd like to push a couple more of extra PRs on cleanup before starting to sketch the API you discussed, in particular wrt the save path.

**Comment by ejgallego** [comment]
The failure is spurious due to a missing overlay from other PR.

**Comment by ejgallego** [comment]
Rebased PR and 4 overlays.

**Comment by ejgallego** [comment]
Ready for merge again.

**Comment by ejgallego** [comment]
@mattam82 it seems Maxime is busy, would you mind taking over this and merging? Thanks!

**Comment by ejgallego** [comment]
By the way @mattam82 I have a patch that integrates the obligation terminators with the main ones, it depends on this and #9129 [so it will take a while] but IMHO it is a nice cleanup and should leave things in good state to attempt for a more flexible API.

**Comment by ejgallego** [comment]
Rebased PR and 3 overlays.

**Comment by ejgallego** [comment]
> Rebased PR and 3 overlays.

Ready for merge again, CI failure is spurious.

**Comment by ejgallego** [comment]
Rebased PR.

**Comment by ejgallego** [comment]
CI green.

**Review by maximedenes — APPROVED** [review]
I'm not sure I like the change of line-breaking style in the interfaces, but let's merge meanwhile and see that later.

**Comment by ejgallego** [comment]
Thanks

**Comment by ejgallego** [comment]
> I'm not sure I like the change of line-breaking style in the interfaces, but let's merge meanwhile and see that later.

I have no strong opinion on it, but it is a style that is much better than the one we have [which is no style and really messy lines], works well for diffs, and it is the one adopted by some other projects including `ocamlformat`, `dune`, etc... so given that it is becoming a short of "standard" in particular a style produced by automatic formatting tools I am OK using it.

**Comment by ejgallego** [comment]
cc: #9572

**Pull request merged**